### PR TITLE
Research and implement support for "ref locals and returns

### PR DESCRIPTION
**Pull request by ivo-stoilov**
o added class for making the default type references
o removed limitation for ref returns
o modified RecursiveMockingBehavior and ImplementationOverrideBehavior behaviors to handle ref return delegates

### CODE REVIEW

File: Telerik.JustMock/Expectations/NonPublicExpectationBase.cs [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+using System.Text;
+using Telerik.JustMock.Core;
+using Telerik.JustMock.Core.Behaviors;
+using Telerik.JustMock.Core.Context;
+using Telerik.JustMock.Expectations.Abstraction;
+using Telerik.JustMock.Expectations.DynaMock;
+
+namespace Telerik.JustMock.Expectations
+{
+    internal abstract class NonPublicExpectationBase
```
Comment (lyubomir-rusev) [inline Telerik.JustMock/Expectations/NonPublicExpectationBase.cs]: This class contains mostly methods that are used as helpers from other parts of the code, like NonPublicExpectation. Based on the functionality in the class I think that it is not very correct to  call this "NonPublicExpectationBase". Also I think that inheriting this class from NonPublicExpectation is not following the "IS A" OOP design rule. 

For example I have also moved this functionality outside NonPublicExpectation when implementing mocking for C# 7 Local Methods. In my case I named the class like "MockingUtil" and used it without inheritance. I think this approach is better in terms of OOP design.

Looking forward for your comments!

File: Telerik.JustMock/Expectations/NonPublicRefReturnExpectation.cs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+   limitations under the License.
+*/
+
+using System;
+using Telerik.JustMock.Core;
+using Telerik.JustMock.Core.Context;
+
+namespace Telerik.JustMock.Expectations.Abstraction
+{
+    internal sealed class NonPublicRefReturnExpectation : NonPublicExpectationBase, INonPublicRefReturnExpectation
```
Comment (lyubomir-rusev) [inline Telerik.JustMock/Expectations/NonPublicRefReturnExpectation.cs]: I add this comment to clearly mark inheritance from NonPublicExpecatationBase which I find to be deviating from "IS A" OOP design principle.

File: Telerik.JustMock/Expectations/NonPublicExpectation.cs [hunk 3]
```diff
@@ -29,211 +29,9 @@ JustMock Lite
 
 namespace Telerik.JustMock.Expectations
 {
-	internal sealed class NonPublicExpectation : INonPublicExpectation
+	internal sealed class NonPublicExpectation : NonPublicExpectationBase, INonPublicExpectation
```
Comment (lyubomir-rusev) [inline Telerik.JustMock/Expectations/NonPublicExpectation.cs]: I add this comment to clearly mark inheritance from NonPublicExpecatationBase which I find to be deviating from "IS A" OOP design principle.

File: Telerik.JustMock/Core/PrivateAccessorBase.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+   limitations under the License.
+*/
+
+using System;
+using System.Linq;
+using System.Reflection;
+
+namespace Telerik.JustMock.Core
+{
+    public abstract class PrivateAccessorBase
```
Comment (lyubomir-rusev) [inline Telerik.JustMock/Core/PrivateAccessorBase.cs]: My comment here will be the same as the one for NonPublicExpecatationBase. Is the functionality in this class enough to call it "PrivateAccessorBase" or is more like enough for a Util class?

File: Telerik.JustMock/Core/PrivateRefReturnAccessor.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+   limitations under the License.
+*/
+
+using System;
+using System.Linq;
+using System.Reflection;
+
+namespace Telerik.JustMock.Core
+{
+    internal class PrivateRefReturnAccessor : PrivateAccessorBase, IPrivateRefReturnAccessor
```
Comment (lyubomir-rusev) [inline Telerik.JustMock/Core/PrivateRefReturnAccessor.cs]: I add this comment to clearly mark inheritance from PrivateAccessorBase which I find to be deviating from "IS A" OOP design principle.

File: Telerik.JustMock/PrivateAccessor.cs [hunk 6]
```diff
@@ -46,10 +45,11 @@ namespace Telerik.JustMock
 	/// dynamic acc = new PrivateAccessor(myobj);
 	/// acc.PrivateProperty = acc.PrivateMethod(123); // PrivateProperty and PrivateMethod are private members on myobj's type.
 	/// </remarks>
-	public sealed class PrivateAccessor : IDynamicMetaObjectProvider
+	public sealed class PrivateAccessor : PrivateAccessorBase, IDynamicMetaObjectProvider
```
Comment (lyubomir-rusev) [inline Telerik.JustMock/PrivateAccessor.cs]: I add this comment to clearly mark inheritance from PrivateAccessorBase which I find to be deviating from "IS A" OOP design principle.

File: Telerik.JustMock/Expectations/NonPublicExpectationBase.cs (hunk 1 above)
Comment (ivo-stoilov) [reply Telerik.JustMock/Expectations/NonPublicExpectationBase.cs]: @lyubomir-rusev, fixed with [595ded1](https://github.com/telerik/JustMockLite/pull/42/commits/595ded10a032d8e33576643110280741ea96a82d)

File: Telerik.JustMock/Expectations/NonPublicRefReturnExpectation.cs (hunk 2 above)
Comment (ivo-stoilov) [reply Telerik.JustMock/Expectations/NonPublicRefReturnExpectation.cs]: @lyubomir-rusev, fixed with [595ded1](https://github.com/telerik/JustMockLite/pull/42/commits/595ded10a032d8e33576643110280741ea96a82d)

File: Telerik.JustMock/Expectations/NonPublicExpectation.cs (hunk 3 above)
Comment (ivo-stoilov) [reply Telerik.JustMock/Expectations/NonPublicExpectation.cs]: @lyubomir-rusev, fixed with [595ded1](https://github.com/telerik/JustMockLite/pull/42/commits/595ded10a032d8e33576643110280741ea96a82d)

File: Telerik.JustMock/Core/PrivateAccessorBase.cs (hunk 4 above)
Comment (ivo-stoilov) [reply Telerik.JustMock/Core/PrivateAccessorBase.cs]: @lyubomir-rusev, fixed with [595ded1](https://github.com/telerik/JustMockLite/pull/42/commits/595ded10a032d8e33576643110280741ea96a82d)

File: Telerik.JustMock/Core/PrivateRefReturnAccessor.cs (hunk 5 above)
Comment (ivo-stoilov) [reply Telerik.JustMock/Core/PrivateRefReturnAccessor.cs]: @lyubomir-rusev, fixed with [595ded1](https://github.com/telerik/JustMockLite/pull/42/commits/595ded10a032d8e33576643110280741ea96a82d)

File: Telerik.JustMock/PrivateAccessor.cs (hunk 6 above)
Comment (ivo-stoilov) [reply Telerik.JustMock/PrivateAccessor.cs]: @lyubomir-rusev, fixed with [595ded1](https://github.com/telerik/JustMockLite/pull/42/commits/595ded10a032d8e33576643110280741ea96a82d)

**Comment by ivo-stoilov** [comment]
@mihail-vladov, @lyubomir-rusev, I have made an important change with [4fe1ab2](https://github.com/telerik/JustMockLite/pull/42/commits/4fe1ab2b8c70e8bf4c1a605daf455dc6c012d864), please review.

File: Telerik.JustMock/Helpers/DynamicTypeHelper.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    /// Constructs dynamic types in memory
+    /// </summary>
+    internal static class DynamicTypeHelper
+    {
+        private const string DynamicAssemblyName = "Telerik.JustMock.DynamicTypes";
+        private static Dictionary<Type, int> typeIndices = new Dictionary<Type, int>();
+
+        public static Type GetNextType<T>()
+        {
+            AssemblyBuilder assemblyBuilder =
```
Comment (mihail-vladov) [inline Telerik.JustMock/Helpers/DynamicTypeHelper.cs]: Can we reuse the defined assembly here?

File: Telerik.JustMock/Helpers/DynamicTypeHelper.cs [hunk 8]
```diff
@@ -0,0 +1,66 @@
+﻿/*
+ JustMock Lite
+ Copyright © 2010-2018 Telerik EAD
```
Comment (mihail-vladov) [inline Telerik.JustMock/Helpers/DynamicTypeHelper.cs]: Could change the copyright notice as it should be 2018 only. This is applicable to all new files.

File: Telerik.JustMock/Core/MockingUtil.CodeGen.cs [hunk 9]
```diff
@@ -173,5 +189,34 @@ public static MethodInfo CreateDynamicMethodWithVisibilityChecks(Type returnType
 			var attr = (DispIdAttribute)Attribute.GetCustomAttribute(member, typeof(DispIdAttribute));
 			return attr != null ? (int?)attr.Value : null;
 		}
+
+		public static ProfilerInterceptor.RefReturn<TReturn> CreateDynamicMethodInvoker<TReturn>(object @object, MethodInfo method, object[] args)
```
Comment (mihail-vladov) [inline Telerik.JustMock/Core/MockingUtil.CodeGen.cs]: Using keywords for variable names does not provide additional context to the variable. Could you change them to have more meaningful names?

File: Telerik.JustMock/Helpers/DynamicTypeHelper.cs (hunk 7 above)
Comment (ivo-stoilov) [reply Telerik.JustMock/Helpers/DynamicTypeHelper.cs]: @mihail-vladov fixed with [41d482f](https://github.com/telerik/JustMockLite/pull/42/commits/41d482f390d2ad05ddea272d25c4b14dca4e999b)

File: Telerik.JustMock/Helpers/DynamicTypeHelper.cs (hunk 8 above)
Comment (ivo-stoilov) [reply Telerik.JustMock/Helpers/DynamicTypeHelper.cs]: @mihail-vladov fixed with [41d482f](https://github.com/telerik/JustMockLite/pull/42/commits/41d482f390d2ad05ddea272d25c4b14dca4e999b)

File: Telerik.JustMock/Core/MockingUtil.CodeGen.cs (hunk 9 above)
Comment (ivo-stoilov) [reply Telerik.JustMock/Core/MockingUtil.CodeGen.cs]: @mihail-vladov fixed with [41d482f](https://github.com/telerik/JustMockLite/pull/42/commits/41d482f390d2ad05ddea272d25c4b14dca4e999b)

**Pull request merged**